### PR TITLE
Add item rest addmetadata endpoint and initial test

### PR DESCRIPTION
**Pull request by mgrauer**
Adds a rest `PUT` endpoint `/item/addmetadata/{id}`.

The request body should be a JSON encoded object with a `metadata` key, containing a list of metadatum.  Each metadatum should have an `element` and `value` key, with corresponding values.  These will be set on the Item metadata.

### CODE REVIEW

File: core/controllers/components/ApiitemComponent.php [hunk 1]
```diff
@@ -171,6 +171,71 @@ public function itemSetmultiplemetadata($args)
...
+        if ($item === false || !$itemModel->policyCheck($item, $userDao, MIDAS_POLICY_WRITE)) {
+            throw new Exception("This item doesn't exist or you don't have write permission.", MIDAS_INVALID_POLICY);
+        }
+        $revisionNumber = array_key_exists('revision', $args) ? (int) $args['revision'] : null;
+        $revision = $apihelperComponent->getItemRevision($item, $revisionNumber);
+
+        if (!array_key_exists(0, $args)) {
+            throw new Exception('Missing request body data.', MIDAS_INVALID_PARAMETER);
+        }
+        $json_body = json_decode($args[0]);
```
Comment (jamiesnape) [inline core/controllers/components/ApiitemComponent.php]: :camel: :baggage_claim:

File: core/controllers/components/ApiitemComponent.php [hunk 2]
```diff
@@ -171,6 +171,71 @@ public function itemSetmultiplemetadata($args)
...
+        if (null === $json_body) {
```
Comment (jamiesnape) [inline core/controllers/components/ApiitemComponent.php]: `$jsonBody === null`

File: core/controllers/components/ApiitemComponent.php [hunk 3]
```diff
@@ -171,6 +171,71 @@ public function itemSetmultiplemetadata($args)
...
+                $item,
+                isset($metadatum->type) ? $metadatum->type : MIDAS_METADATA_TEXT,
+                $metadatum->element,
+                isset($metadatum->qualifier) ? $metadatum->qualifier : '',
+                $metadatum->value,
+                $revision
+            );
+        }
+
+        return true;
```
Comment (jamiesnape) [inline core/controllers/components/ApiitemComponent.php]: Would it be nicer to return the updated item?

File: core/controllers/components/ApiitemComponent.php [hunk 4]
```diff
@@ -171,6 +171,71 @@ public function itemSetmultiplemetadata($args)
...
+        foreach ($metadata as $metadatum) {
+            if (!isset($metadatum->element) || !isset($metadatum->value)) {
+                throw new Exception("All metadata must have 'element' and 'value' keys.", MIDAS_INVALID_PARAMETER);
+            }
+        }
+
+        foreach ($metadata as $metadatum) {
+            $apihelperComponent->setMetadata(
+                $item,
+                isset($metadatum->type) ? $metadatum->type : MIDAS_METADATA_TEXT,
```
Comment (jamiesnape) [inline core/controllers/components/ApiitemComponent.php]: No need to change, but I note the absence of :chicken:-typing.

File: core/tests/controllers/api/CMakeLists.txt [hunk 5]
```diff
@@ -19,3 +19,4 @@
 
 add_midas_test(CoreRestCallUserMethodsTest RestCallUserMethodsTest.php)
 add_midas_test(CoreRestKeyControllerTest RestKeyControllerTest.php)
+add_midas_test(CoreRestCallItemMethodsTest RestCallItemMethodsTest.php)
```
Comment (jamiesnape) [inline core/tests/controllers/api/CMakeLists.txt]: Can you alphabetize these by test name?

File: core/tests/controllers/api/RestCallItemMethodsTest.php [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+ See the License for the specific language governing permissions and
+ limitations under the License.
+=========================================================================*/
+
+require_once BASE_PATH.'/core/tests/controllers/api/RestCallMethodsTestCase.php';
+
+/** Tests the functionality of the web API Rest Item methods */
+class Core_RestCallItemMethodsTest extends RestCallMethodsTestCase
+{
+    /** set up tests */
```
Comment (jamiesnape) [inline core/tests/controllers/api/RestCallItemMethodsTest.php]: Set up tests.

File: core/tests/controllers/api/RestCallItemMethodsTest.php [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+=========================================================================*/
+
+require_once BASE_PATH.'/core/tests/controllers/api/RestCallMethodsTestCase.php';
+
+/** Tests the functionality of the web API Rest Item methods */
```
Comment (jamiesnape) [inline core/tests/controllers/api/RestCallItemMethodsTest.php]: methods.

File: core/tests/controllers/api/RestCallItemMethodsTest.php [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+        parent::setUp();
+    }
+
+    /** Set metadata on the item. */
+    public function testItemAddmetadata()
+    {
+        $itemsFile = $this->loadData('Item', 'default');
+        $itemDao = $this->Item->load($itemsFile[1]->getKey());
+
+        $apipath = '/item/addmetadata/'.$itemDao->getItemId();
```
Comment (jamiesnape) [inline core/tests/controllers/api/RestCallItemMethodsTest.php]: $apiPath

**Comment by jamiesnape** [comment]
Does this add metadata or overwrite it?

**Comment by mgrauer** [comment]
>Does this add metadata or overwrite it?

a bit of both, it's an upsert.  name suggestions?

**Comment by jamiesnape** [comment]
In an ideal world it would be `PATCH /item;metadata` or `PUT /item;metadata`. `PUT /item/addmetadata` is as good a name as Midas has ever had for anything, so it is fine.

**Comment by jamiesnape** [comment]
A few minor style points, but otherwise LGTM.

**Comment by mgrauer** [comment]
@jamiesnape Feedback addressed.

**Comment by jamiesnape** [comment]
LGTM.

**Pull request merged**